sealtk/core/VideoSource.hpp

```diff
@@ -23,44 +18,113 @@ namespace sealtk
 class SEALTK_CORE_EXPORT VideoSource : public QObject
 {
   Q_OBJECT
 
 public:
-  explicit VideoSource(QObject* parent = nullptr);
   ~VideoSource() override;
 
+  /// Query if the video source is "ready".
+  ///
+  /// This method is used to determine if the video source is "ready". In
+  /// particular, this method can be used to determine if #frames will return
+  /// meaningful data. A source will not return \c true until #start has been
+  /// called, and until the initial frame set has been passed from the source's
+  /// internal thread to the thread which owns the #VideoSource object.
+  virtual bool isReady() const = 0;
+
+  /// Get the set of frames for which this source has video.
+  ///
+  /// This method returns a map of times for which the video source has video,
+  /// mapped to their corresponding frames. Note that calling this method
+  /// before #framesChanged() is emitted may return an empty map.
   virtual TimeMap<kwiver::vital::timestamp::frame_t> frames() const = 0;
 
 signals:
-  void imageReady(
-    kwiver::vital::image_container_sptr const& image,
-    VideoMetaData const& metaData) const;
-  void detectionsReady(
-    kwiver::vital::detected_object_set_sptr const& detetedObjectSet,
-    kwiver::vital::timestamp const& timeStamp) const;
-  void videoInputChanged() const;
+  /// Emitted when the set of available frames changes.
+  ///
+  /// This signal is emitted when the set of available frames changes. While
+  /// this has obvious use for "streaming" sources (that is, the set of frames
+  /// is properly time variable), even "static" sources, because video is
+  /// normally loaded asynchronously, should emit this signal at least once to
+  /// indicate when they are ready for users to call #frames.
+  void framesChanged();
 
 public slots:
-  virtual void invalidate() const = 0;
-  virtual void seek(kwiver::vital::timestamp::time_t time, SeekMode mode) = 0;
-  virtual void seekFrame(kwiver::vital::timestamp::frame_t frame) = 0;
-
-  virtual void seekTime(kwiver::vital::timestamp::time_t time)
-  { seek(time, SeekExact); }
+  /// "Start" the video source.
+  ///
+  /// This method finalizes the construction of a video source. It must be
+  /// called prior to using the video source, preferably by whatever code
+  /// constructed the video source. If necessary, it will be called by
+  /// #requestFrame, however this may impose a non-trivial latency on initial
+  /// requests if the source must perform work in the source thread to prepare
+  /// the source for use.
+  ///
+  /// The video source will not emit signals until this method has been called.
+  /// If necessary, this gives users a chance to connect to #framesChanged
+  ///
+  /// \note
+  ///   It is safe to call this method more than once, as long as it is not
+  ///   called during destruction of the #VideoSource. Once this method is
+  ///   called, the underlying #VideoProvider is moved to the video source's
+  ///   service thread.
+  void start();
```

Comment:
How is this going to be handled if/when we port this to WASM?

Reply:
Uh... hopefully it will Just Work? In any case, we definitely need the asynchronous stuff for WASM...

Reply:
Web doesn't have threads. How will it "just work"?

Reply:
A brief Google search seems to imply that at least Chrome ≥ 70 does...